crates/ark/src/lsp/completions/function_context.rs

```diff
@@ -0,0 +1,228 @@
+
+        let usage = determine_function_usage(
+            &effective_function_node,
+            &document_context.document.contents,
+        );
+
+        let function_name_node = if effective_function_node.is_namespace_operator() {
+            // Note: this could be 'None', in the case of, e.g., `dplyr::@`
+            effective_function_node.child_by_field_name("rhs")
+        } else {
+            Some(effective_function_node)
+        };
+
+        let name = match function_name_node {
+            Some(node) => node_text(&node, &document_context.document.contents).unwrap_or_default(),
+            None => String::new(),
+        };
+
+        let arguments_status = if usage == FunctionUsage::Reference {
+            ArgumentsStatus::Absent
+        } else {
+            determine_arguments_status(&effective_function_node)
+        };
+
+        log::info!(
+            "FunctionContext created with name: '{name}', usage: {usage:?}, arguments: {arguments_status:?}"
+        );
+
+        Self {
+            name,
+            range: match function_name_node {
+                Some(node) => convert_tree_sitter_range_to_lsp_range(
+                    &document_context.document.contents,
+                    node.range(),
+                ),
+                None => {
+                    // Create a zero-width range at the end of the effective_function_node
+                    let node_range = effective_function_node.range();
+                    let end_position = convert_tree_sitter_range_to_lsp_range(
+                        &document_context.document.contents,
+                        node_range,
+                    )
+                    .end;
+                    tower_lsp::lsp_types::Range::new(end_position, end_position)
+                },
+            },
+            usage,
+            arguments_status,
+        }
+    }
+}
+
+/// The practical definition of the effective function node is "Which node
+/// should I take the parent of, if I want the parent of a function call or
+/// reference?"
+///
+/// This handles both simple identifiers (`fnc`) and namespace-qualified
+/// identifiers (`pkg::fcn`).
+///
+/// The alleged function node has to either be an identifier or a
+/// namespace operator. Otherwise, we return `None`.
+fn get_effective_function_node(node: Node) -> Option<Node> {
+    let Some(parent) = node.parent() else {
+        return None;
+    };
+
+    if parent.is_namespace_operator() {
+        Some(parent)
+    } else if node.is_identifier() || node.is_namespace_operator() {
+        Some(node)
+    } else {
+        None
+    }
+}
+
+/// When completing a function inside these functions, we treat it as a value
+/// reference (don't automatically add parentheses).
+static FUNCTIONS_EXPECTING_A_FUNCTION_REFERENCE: &[&str] = &[
+    "args",
+    "debug",
+    "debugonce",
+    "formals",
+    "help",
+    "trace",
+    "str",
+];
+
+fn is_inside_special_function(node: &Node, contents: &ropey::Rope) -> bool {
```

Comment:
Can we find a more explicit name?

`is_expected_function_value()` or something like that?